### PR TITLE
feat(popover): add the pf popover

### LINKED ISSUE #20: Component - Popover
Implement a component that provides the following UI element:

- [PatternFly Popover](http://www.patternfly.org/pattern-library/widgets/#popover)

Acceptance Criteria:

- When rendered, the component matches the examples in the PatternFly Pattern Library (see link above)
- the component supports:
  - displaying a title
  - displaying a close icon

### PR DESCRIPTION
**Pull request by dabeng**
**What**:
This PR adds the PatternFly popover patterns. To be used to close #20 (Popover) once done.

Storybook: https://rawgit.com/dabeng/patternfly-react/popover-storybook/.out/index.html?knob-Label=Danger%20Will%20Robinson%21&knob-Popover%20Title=Popover&knob-Popover%20Content=%3Cstrong%3EHoly%20guacamole%21%3C%2Fstrong%3E%20Check%20this%20info.&knob-Placement=right&knob-Trigger=click&knob-Root%20Close=true&selectedKind=Popover&selectedStory=Popover&full=0&down=1&left=1&panelRight=0&downPanel=storybooks%2Fstorybook-addon-knobs

**How**:
Encapsulating react-bootstrap popover as a story of patternfly-react storybook.

@priley86 , please review.

### CODE REVIEW

**Comment by sharvit** [comment]
@dabeng same as my comment about the Tooltip:
https://github.com/patternfly/patternfly-react/pull/79#issuecomment-346555496

**Comment by dabeng** [comment]
Hi @sharvit , I have refactored PR according to your advice.

File: src/Popover/Popover.stories.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    <Popover id="popover" title="Popover">
+      <strong>Holy guacamole!</strong> Check this info.
+    </Popover>
+  )
+  const placement = select(
+    'Placement',
+    ['top', 'bottom', 'left', 'right'],
+    'right'
+  )
+  const trigger = text('Trigger', 'hover focus')
```
Comment (priley86) [inline src/Popover/Popover.stories.js]: I would use a select knob here w/ options `click`, `hover`, and `focus` instead of a text knob.

**Comment by priley86** [comment]
@dabeng nice job w/ this. Two other things I'm noticing here:

* The only [documentation](https://github.com/patternfly/patternfly-design/blob/master/pattern-library/forms-and-controls/help-on-forms/design/design.md) I'm currently seeing on popover states that it should close when the user clicks outside the overlay. For that reason I think we should set the `rootClose` prop on the Overlay to true. Our other existing [examples](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/popovers.html) of popovers have close icons, but I don't think that is necessarily a requirement in the current design docs. It seems that adding the close icon will require a customized overlay and may be additional work, so would we be OK w/ moving forward w/o this for now @mcarrano ?

* The popover title text and popover text would be nice to include a text knobs i.m.o.

**Comment by mcarrano** [comment]
@dabeng @priley86 Looks like this is behaving more like a tooltip. In the examples on PatternFly.org the popover only appears on click, not hover, although the documentation is not explicit. Does this make sense to you?

**Comment by priley86** [comment]
@mcarrano @dabeng It looks like the API surface is very similar for Tooltips and Popovers w/ React Bootstrap. I am seeing that we have the following triggers w/ this one:
https://react-bootstrap.github.io/components.html#popovers-trigger-behaviors 

* _Click_ - click button to open, click button again to close
* _Hover + focus_ (or just hover but typically you'd include focus) - hover button w/ mouse or focus it with tab key to open
* _Focus_ - focus w/ tab key or click to open
* _Click w/ root close_ - click button to open and click anywhere else to close

Based on your feedback @mcarrano, it sounds we can use _click w/ root close_ for popover examples and _hover + focus_ for tooltip examples? This API would be exposed to the consumer (so they'd have flexibility), but it's probably best to demonstrate the common use cases that we'd expect. We can probably circle back around on this later when we get into forms and other layouts w/ tooltips and popovers, but just hoping to provide some general examples now.

@mcarrano @dabeng any more thoughts on this?

**Comment by dabeng** [comment]
Hi @priley86 @mcarrano  , I have refactored the PR according to your advices. The changes are listed below:
1. Allow users to modify the popover title text and content text
2. Users can only trigger popover by clicking the target button
3. add editor for prop roorClose
4. It's hard to add "custom overlay" example to current storybook due to this issue -- [ Stateless function components cannot be given refs](https://github.com/facebook/react/issues/10831)

**Comment by dabeng** [comment]
Hi @priley86 @mcarrano  , I have refactored the PR according to your advices. The changes are listed below:
1. Allow users to modify the popover title text and content text
2. Clicking the target button is the default trigger action for popover
3. add editor for prop roorClose
4. It's hard to add "custom overlay" example to current storybook due to this issue -- [ Stateless function components cannot be given refs](https://github.com/facebook/react/issues/10831)

**Comment by mcarrano** [comment]
Yes, @priley86 "click w/ root close" seems to match the popover behavior that we expect. I know that there was also some question about when and if an explicit Close button can be added within the popover. I think this is a nice option to have, but not critical IMO. I think that most users will just click away from the popover if they don't see an explicit way to close it.

**Comment by dabeng** [comment]
Hi @priley86 @jgiardino . Great solution. I made the following adjustments:
```js
   const trigger = select(
     'Trigger',
     ['hover', 'focus', 'hover focus', 'click'],
     'click'
   )
```

**Comment by priley86** [comment]
last thing here... commit should be titled:

`feat(popover): add the pf popover`

instead of:
`WIP:feat(popover):add the pf popover`

**Comment by priley86** [comment]
I would like to go ahead and merge this PR so that can we provide the base functionality for popover. We can try to add support the close icon in subsequent PRs.

@jgiardino @mcarrano thanks very much for the reviews. Can you please follow up in this issue if there is anything else? Or close this issue and create a new one?
https://github.com/patternfly/patternfly-react/issues/20

**Pull request merged**